### PR TITLE
Fix filtering for downloadable blocks in inserter.

### LINKED ISSUE #55209: Hiding blocks breaks block directory
### Description

As soon as you hide at least one block in the settings the block directory doesn't display any results.

The results get still returned from the REST API so the problem seems to be client side in the block editor.

### Step-by-step reproduction instructions

1. Open a post
2. Open the block inserter and search for something that yields a block directory result, e.g. "Sharing"
3. See the blocks available to install
4. Open Preferences
5. Open the "Blocks" panel
6. Disable any block
7. Close the Preferences
8. Open the block inserter and again search for something that yields a block directory results, e.g. "Sharing"
9. No blocks to install are displayed

### Screenshots, screen recording, code snippet

[Bildschirmaufzeichnung vom 2023-10-10, 16:47:21.webm](https://github.com/WordPress/gutenberg/assets/2428517/186c8de8-b655-45f2-b282-e3c104bdca69)

### Environment info

- WordPress 6.3.1
- WordPress 6.3.1 with Gutenberg 16.7.1 active
- Desktop with Firefox 118.0.1 as well as Chrome 117.0.5938.149

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

Fixes #55209.

Changes how the `downloadableItems` passed to `DownloadableBlocksPanel` are calculated so that if blocks are hidden in the post editor a list of downloadable blocks still shows.

The problem was we were using `canInsertBlockType` to check if the downloadable blocks could be used, but that function only checks against a list of blocks that are already installed. A block that's not installed won't be on the list, so it would always return false. 

The reason this logic magically worked if there were no blocks hidden in the editor is that `allowedBlockTypes` is a boolean if there are no _disallowed_ block types, but a list if there are 😅 

I changed the logic to filter out blocks already installed, because I noticed installed blocks were still appearing in the downloadable blocks results and if a block is installed but hidden, we don't want it appearing there. I may be missing something here though - is there any reason we might want installed blocks to still appear on the list? If so, we could add logic to check the installed blocks against `allowedBlockTypes` instead.

## Testing Instructions

1. In the post editor, search for a term in the inserter, e.g. "Searching";
2. Note the results that appear;
3. Go to Options > Preferences > Blocks and hide a couple of blocks;
4. Search the inserter again and check the same results appear;
5. Install one of the blocks;
6. In Options > Preferences > Blocks hide the block just installed;
7. Search again and check that installed block doesn't appear.

With MathML block installed and hidden:

#### before

Screenshot 2023-10-11 at 3 11 17 pm

#### after

Screenshot 2023-10-11 at 3 45 09 pm

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
Oh, I like this idea in principle! Once a block is installed, you likely don't need to see the "Available to install" UI 👍

One potential wrinkle is that when installing a block, it now hides the area as soon as the block is installed, which could feel a bit jumpy. With a single block it looks kinda okay:

https://github.com/WordPress/gutenberg/assets/14988353/1b5f6f74-96ad-4e28-969c-1c725715f5f0

However, if the list is fairly long or if the entered string to install the block doesn't quite match the target block, it's possible to wind up in a state where the installed block doesn't appear to be visible anywhere in the list, and focus appears to now be nowhere in particular:

https://github.com/WordPress/gutenberg/assets/14988353/587de4b9-d255-4f49-90a1-77b97bc8eb10

A side note on the above is that in that example the results aren't very helpful based on the search term, but that's a separate issue! 😅

> If so, we could add logic to check the installed blocks against allowedBlockTypes instead.

Given the potential jumpiness during the installation flow, maybe checking against `allowedBlockTypes` would make the UI feel a bit more stable there?

**Comment by tellthemachines** [comment]
Thanks for testing @andrewserong !

> Given the potential jumpiness during the installation flow, maybe checking against allowedBlockTypes would make the UI feel a bit more stable there?

I've fixed this to check if the block has just been installed, in which case we should show it.

I also did a bit more digging into the reason for the initial logic and found #25926, which indicates that `allowedBlockTypes` was being used to check if we're trying to insert a block in a context that has an allow list, such as the Navigation block, or the Columns block. This complicates matters a bit 😅 

A possible solution is to try and differentiate between the allow list and the disallow list scenarios (disallow list should be what happens when we hide blocks in preferences)

On the other hand, should it be possible to insert custom blocks targeted specifically at blocks such as Navigation? I half remember this being discussed at some point but I'm not sure it was ever worked on.

**Comment by andrewserong** [comment]
Oh, interesting problem!

> On the other hand, should it be possible to insert custom blocks targeted specifically at blocks such as Navigation? I half remember this being discussed at some point but I'm not sure it was ever worked on.

If so, I imagine that'd be the responsibility of the plugin to figure out, rather than for us to handle in the logic here? Given that #25926 landed a long time ago now, I suppose it might be worth seeing if we can honour the intentions behind that PR. I guess what I mean is, if a plugin decides to override core blocks' allow lists, then I think it'd probably be fine for that behaviour to happen within the plugin, rather than need to account for it in this block directory listing... I don't feel strongly about it, though! 🙂

> A possible solution is to try and differentiate between the allow list and the disallow list scenarios (disallow list should be what happens when we hide blocks in preferences)

Is one challenge there that the disallow list happens over in the post editor store, rather than with the block directory package?

---

All that said, personally I don't mind the approach of this PR. Would it be worth getting a second opinion from some of the folks who worked on #25926 originally, to get a better sense of desired behaviour nowadays?

**Comment by andrewserong** [comment]
Apologies, I meant to add — the latest commit tests very well for me according to the testing instructions, and resolves the issue of the installed state disappearing. The UI for the installed state is working nicely for the rest of the current session in the editor, and after a reload, if you search for a block, you just see that block rather than the listing for the plugin:

image

So, all up, this LGTM if we feel confident that it's okay that we no longer do the check for allowed blocks!

**Comment by tellthemachines** [comment]
> All that said, personally I don't mind the approach of this PR. Would it be worth getting a second opinion from some of the folks who worked on #25926 originally, to get a better sense of desired behaviour nowadays?

That's why I added @ryelle to the reviewers 😄 though perhaps @StevenDufresne might have thoughts on this too?

I dunno, it's definitely not a great experience to try to install/add a custom block to, say, the Navigation block. On one hand there aren't that many blocks that limit the types of blocks that can be added without just inserting one block type automatically, so this not-so-great experience may not be as common as trying to install blocks in the post editor while also having hidden blocks. If that's the case, this change should be OK. 

I've been looking for how to tell that the allow list comes from a parent block instead of the post editor, but there doesn't seem to be a way of doing so. The only alternative I can think of is checking the block type of the root client id, which should be undefined if we're at root level in the post editor. This would only allow us to fix the issue of installing new blocks at root level though - it still wouldn't work within a Group block.

**Comment by andrewserong** [comment]
> I dunno, it's definitely not a great experience to try to install/add a custom block to, say, the Navigation block.

Good point... I suppose one of the reasons I don't mind this PR so much is that a user can install the block via the plugin, and it (mostly) gracefully falls back to installing, but not inserting the block where it can't go. For example, if I have a child block of the Navigation block selected when I go to add MathML, the block gets installed, it just doesn't get inserted anywhere:

https://github.com/WordPress/gutenberg/assets/14988353/5d9f1dbf-fa34-4614-b736-4eea9d60c256

If we go with this PR, one tiny improvement there could be to update the snackbar notice if it wasn't able to insert so that it just says `Block %s installed` rather than `Block %s installed and added`?

**Review by ramonjd — APPROVED** [review]
This is huge improvement on trunk, so leaving a LGTM pending further feedback/suggestions for follow ups from pinged people.

**Comment by tellthemachines** [comment]
Thanks for the feedback folks! I'll let it sit for a day or two to give others a chance to say something. This isn't for 6.4 so no huge rush 😄

**Comment by StevenDufresne** [comment]
@tellthemachines I can't add any more context that isn't shared in the two issues linked in the PR (https://github.com/WordPress/gutenberg/pull/22240 & https://github.com/WordPress/gutenberg/pull/24148). Regarding the issue of allowed lists and inserting blocks, I don't know how likely it is. When this was written we weren't sure how the block directory would be used since there were ideas of only accepting single block plugins (which could lead to separate child block plugins). 

I would run through the test case referenced in https://github.com/WordPress/gutenberg/pull/24148 for safety. Short of that, you are now the expert 🥇  :)

**Comment by tellthemachines** [comment]
> I would run through the test case referenced in #24148 for safety.

Thanks for the tip @StevenDufresne ! The test case in that PR is working correctly on this branch, so should be good from that perspective.

> If we go with this PR, one tiny improvement there could be to update the snackbar notice if it wasn't able to insert so that it just says Block %s installed rather than Block %s installed and added?

Hmm, I can't see any obvious way of doing this given that we're unable to tell whether the block is allowed or not, and I don't think there's a way to tell whether it has been successfully added to the page. 

I'm inclined to merge this as is and revisit/iterate based on any feedback that comes in afterwards.

**Comment by andrewserong** [comment]
> I'm inclined to merge this as is and revisit/iterate based on any feedback that comes in afterwards.

Me, too. It's quite a niche / edge case situation, so a plus +1 from me for merging and updating further down the track if anyone winds up finding it confusing.

**Pull request merged**
